File: problems/problemset/virtual-filesystem/Python/runner/testbench.py
import sys
import json
import time
import resource

sys.path.insert(0, '/workspace/code')
from vfs import VFS

def main():
    testcase = json.loads(sys.stdin.read())
    vfs = VFS()
    output = ''

    start = time.perf_counter()

    for line in testcase['input'].splitlines():
        if not line:
            continue
        parts = line.split()
        cmd = parts[0]

        if cmd == 'MKDIR':
            output += vfs.mkdir(parts[1]) or ''
        elif cmd == 'TOUCH':
            output += vfs.touch(parts[1]) or ''
        elif cmd == 'WRITE':
            _, path, *rest = line.split(' ', 2)
            content = rest[0] if rest else ''
            output += vfs.write(path, content) or ''
        elif cmd == 'READ':
            output += vfs.read(parts[1]) or ''
        elif cmd == 'LS':
            output += vfs.ls(parts[1]) or ''
        elif cmd == 'RM':
            if parts[1] == '-r':
                output += vfs.rmr(parts[2]) or ''
            else:
                output += vfs.rm(parts[1]) or ''
        elif cmd == 'MV':
            output += vfs.mv(parts[1], parts[2]) or ''
        elif cmd == 'CP':
            output += vfs.cp(parts[1], parts[2]) or ''
        elif cmd == 'FIND':
            output += vfs.find(parts[1], parts[2]) or ''

    elapsed_ms = (time.perf_counter() - start) * 1000
    memory_kb = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
    result = {'actual_output': output, 'time_ms': elapsed_ms, 'memory_kb': memory_kb}
    sys.stdout.write(f'__BEGIN_RESULT__{json.dumps(result)}__END_RESULT__')

main()

File: problems/problemset/virtual-filesystem/C++/runner/Testbench.cpp
#include <bits/stdc++.h>
#include <nlohmann/json.hpp>
#include <chrono>
#include <sys/resource.h>
#include "../code/VFS.h"

using json = nlohmann::json;

static long peakMemoryKB() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss;
}

int main() {
    std::string raw((std::istreambuf_iterator<char>(std::cin)),
                     std::istreambuf_iterator<char>());
    json testcase = json::parse(raw);
    auto input = testcase["input"].get<std::string>();

    std::stringstream ss(input);
    std::string line;
    std::string output;
    VFS vfs;

    auto start = std::chrono::high_resolution_clock::now();

    while (std::getline(ss, line)) {
        if (line.empty()) continue;
        std::istringstream tok(line);
        std::string cmd;
        tok >> cmd;

        if (cmd == "MKDIR") {
            std::string path; tok >> path;
            output += vfs.mkdir(path);
        } else if (cmd == "TOUCH") {
            std::string path; tok >> path;
            output += vfs.touch(path);
        } else if (cmd == "WRITE") {
            std::string path; tok >> path;
            std::string content;
            std::getline(tok, content);
            if (!content.empty() && content[0] == ' ') content = content.substr(1);
            output += vfs.write(path, content);
        } else if (cmd == "READ") {
            std::string path; tok >> path;
            output += vfs.read(path);
        } else if (cmd == "LS") {
            std::string path; tok >> path;
            output += vfs.ls(path);
        } else if (cmd == "RM") {
            std::string next; tok >> next;
            if (next == "-r") {
                std::string path; tok >> path;
                output += vfs.rmr(path);
            } else {
                output += vfs.rm(next);
            }
        } else if (cmd == "MV") {
            std::string src, dst; tok >> src >> dst;
            output += vfs.mv(src, dst);
        } else if (cmd == "CP") {
            std::string src, dst; tok >> src >> dst;
            output += vfs.cp(src, dst);
        } else if (cmd == "FIND") {
            std::string path, name; tok >> path >> name;
            output += vfs.find(path, name);
        }
    }

    auto end = std::chrono::high_resolution_clock::now();
    double time_ms   = std::chrono::duration<double, std::milli>(end - start).count();
    long   memory_kb = peakMemoryKB();

    json result;
    result["actual_output"] = output;
    result["time_ms"]       = time_ms;
    result["memory_kb"]     = memory_kb;

    std::cout << "__BEGIN_RESULT__" << result.dump() << "__END_RESULT__";
}
